### DISCUSSION TITLE
Behaviour of get_node

### DISCUSSION THREAD

**Discussion by Randelung** [FreeOpcUa/opcua-asyncio#523] [Q&A] [answered]
Hi, I was wondering about the behaviour of `get_node` and its stability.
Right now you can `get_node` any arbitrary ID and get a `Node` object, without any checks whether the node exists or not. You only discover that it doesn't when you try to interact with it.
I've inadvertently used that mechanic in the setup of some data binding in my program. The nodes are being `get`ed before they exist on the server and a callback that writes to the node is set up.
After, the nodes are actually being created, and it's only then that the nodes will be interacted with.

I'm trying to gauge if that might change in the future and possibly break that setup, as it's possible you might include a check.
Do you have any plans of modifying this behaviour?

Thanks!

**Comment by AndreasHeine** [comment, marked as answer]
actually the node class is just a gathering of usefull highlevel methods otherwise you need to make lowlevel calls (like in uaclient). usually in my clients i try to avoid the nodeclass as much as i can because its one request for just one part of information (value, browsename, ...). i prefer to bundle those request and read multiple nodes attributes what so ever in just one request to reduce the network traffic (but check the servers maxnodesperbrowse, etc up front^^)! if you trace your opc ua application you will see a lot of unnecessary requests (at least my oppinion)

a checkup from serverside wouldnt be so costly but from client side... it would be a request to verify the node exist.

**Reply by Randelung** [reply]
I'm not querying a lot, anyway. I have an application that is server in one direction and about 17 clients in the other. I'm subscribed to data changes in the clients and write to server nodes, so the network overhead should really be minimal. I'm only actually querying the service level node to check for connection problems (per your recommendation :) ).

But I gather from your answer that you don't intend to introduce checking for existence?

**Reply by AndreasHeine** [reply]
its just my personal oppinion!

if someone has a good idea... i am open minded for other solutions ;)

**Reply by AndreasHeine** [reply]
maybe on serveside writing is a better solution
https://github.com/FreeOpcUa/opcua-asyncio/blob/80a4cc2635e8888ddda051f6dc282e2873570d0e/asyncua/server/server.py#L666

**Reply by Randelung** [reply]
Ah I see, you want me to circumvent the possibility of a breaking change in the future altogether by using this instead of a node object. I'll have a look into this tomorrow, right now I'm about to fall off my chair.
Thanks for your input.

**Reply by oroulet** [reply]
I use node objects a lot. I think it really depends on the applications. In my use cases there is usually no extra calls.
And no there is no plan to change the behavior of get_node ()

**Reply by Randelung** [reply]
OK cool, thanks!

**Reply by jmarshall9120** [reply]
@AndreasHeine - Could perhaps provide and example on how your query multiple nodes in one request?  I haven't found any examples of that.

**Reply by AndreasHeine** [reply]
> @AndreasHeine - Could perhaps provide and example on how your query multiple nodes in one request?  I haven't found any examples of that.

 https://github.com/AndreasHeine/opcua-tutorial/blob/da44b65682ff70ace7feb640d6d43c7d2880f7ed/client/read.py#L22

**Reply by jmarshall9120** [reply]
@AndreasHeine 

That doc you linked is an awesome resource!  One question though, are we supposed to assume that the results are synchronous when using `client.uaclient.read_attributes`?

I noticed the `result` object (I was testing with `DataValue` results by passing `ua.AttributeIds.Value`) doesn't return any information linking it to the NodeId I passed.  Due to this I would have to assume the data returned in the same order as the list of NodeId passed to `client.ua.read_attributes`.  This is an assumption one would never make on something like an SQL server.

**Reply by AndreasHeine** [reply]
opc ua spec. says that it has to be in the same order like the requested nodeids so the result idx matches
